tools/dkcm/mainservice/mainservice.go

```diff
@@ -0,0 +1,236 @@
+	ClusterInfo *clerk.Response `json:"clusterInfo"`
+}
+
+func Start(dbConfig *mysql.DBConfig, boskosClientHost, gcpServiceAccount string) error {
+	var err error
+	boskosClient, err = boskos.NewClient(boskosClientHost, "", "")
+	if err != nil {
+		return fmt.Errorf("failed to create Boskos client: %w", err)
+	}
+	dbClient, err = clerk.NewDB(dbConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create Clerk client: %w", err)
+	}
+	serviceAccount = gcpServiceAccount
+	server := http.NewServeMux()
+	server.HandleFunc("/request-cluster", handleNewClusterRequest)
+	server.HandleFunc("/get-cluster", handleGetCluster)
+	server.HandleFunc("/clean-cluster", handleCleanCluster)
+	// use PORT environment variable, or default to 8080
+	port := DefaultPort
+	if fromEnv := os.Getenv("PORT"); fromEnv != "" {
+		port = fromEnv
+	}
+	// start the web server on port and accept requests
+	log.Printf("Server listening on port %q", port)
+	return http.ListenAndServe(fmt.Sprintf(":%v", port), server)
+}
+
+// handle cleaning cluster request after usage
+func handleCleanCluster(w http.ResponseWriter, req *http.Request) {
+	// add project name
+	token := req.URL.Query().Get("token")
+	r, err := dbClient.GetRequest(token)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("there is an error getting the request with the token: %v, please try again", err), http.StatusForbidden)
+		return
+	}
+	c, err := dbClient.GetCluster(r.ClusterID)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("there is an error getting the cluster with the token: %v, please try again", err), http.StatusForbidden)
+		return
+	}
+	err = dbClient.DeleteCluster(r.ClusterID)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("there is an error deleting the cluster with the token: %v, please try again", err), http.StatusForbidden)
+		return
+	}
+	err = boskosClient.ReleaseGKEProject(c.ProjectID)
+	if err != nil {
+		http.Error(w, "there is an error releasing Boskos's project. Please try again.", http.StatusInternalServerError)
+		return
+	}
+}
+
+// check the pool capacity and create clusters if necessary
+func checkPoolCap(cp *clerk.ClusterParams) {
+	chanLock <- struct{}{}
+	numAvail := dbClient.CheckNumStatus(cp, Ready)
+	numWIP := dbClient.CheckNumStatus(cp, WIP)
+	diff := DefaultOverProvision - numAvail - numWIP
+	var wg sync.WaitGroup
+	// create cluster if not meeting overprovisioning criteria
+	for i := int64(0); i < diff; i++ {
+		wg.Add(1)
+		log.Printf("Creating a new cluster: %v", cp)
+		go CreateCluster(cp, &wg)
+	}
+	wg.Wait()
+	<-chanLock
+}
+
+// assign clusters if available upon request
+func CreateCluster(cp *clerk.ClusterParams, wg *sync.WaitGroup) {
+	project, err := boskosClient.AcquireGKEProject(boskos.GKEProjectResource)
+	if err != nil {
+		log.Printf("Failed to acquire a project from boskos: %v", err)
+		return
+	}
+	projectName := project.Name
+	c := clerk.NewCluster(clerk.AddProjectID(projectName))
+	c.ClusterParams = cp
+	clusterID, err := dbClient.InsertCluster(c)
+	wg.Done()
```

Comment:
This should be changed to `defer wg.Done()` and moved to the top of this function.